src/Matrix.jl

```diff
@@ -6725,16 +6725,7 @@ end
 
 Return the $r \times c$ zero matrix over $R$.
 """
-function zero_matrix(R::NCRing, r::Int, c::Int)
-   arr = Matrix{elem_type(R)}(undef, r, c)
-   for i in 1:r
-      for j in 1:c
-         arr[i, j] = zero(R)
-      end
-   end
-   z = Generic.MatSpaceElem{elem_type(R)}(R, arr)
-   return z
-end
+zero_matrix(R::NCRing, r::Int, c::Int) = zero!(dense_matrix_type(R)(R, undef, r, c))
```

Comment:
Nice. This will allow some code deduplication in Nemo, where something like this is currently implemented for every Matrix type

Reply:
Well, not quite: there we often will want to *omit* the `zero!` because the "undef" matrices are actually already zero initialized. E.g. for `ZZMatrix`.

So they'd still have to have a method like
```julia
zero_matrix(R::ZZRing, r::Int, c::Int) = (dense_matrix_type(R)(R, undef, r, c))
```
or maybe rather
```julia
zero_matrix(R::ZZRing, r::Int, c::Int) = ZZMatrix(r, c)
```


However this made me realize a problem: we don't require the `UndefInitializer` constructors to perform a sanity check on `r` and `c` (i.e. that they are `>= 0`). And we probably don't want to (it is not needed when called from `similar`).

So this method here should actually have such a check. But that's fine. And then we'd keep the existing `zero_matrix` method for `ZZRing` as it is. However, we could most or all of these methods from `fmpz_mat.jl:
```
function (a::ZZMatrixSpace)()
  z = ZZMatrix(nrows(a), ncols(a))
  return z
end

function (a::ZZMatrixSpace)(arr::AbstractMatrix{T}) where {T <: IntegerUnion}
  _check_dim(nrows(a), ncols(a), arr)
  z = ZZMatrix(nrows(a), ncols(a), arr)
  return z
end

function (a::ZZMatrixSpace)(arr::AbstractVector{T}) where {T <: IntegerUnion}
  _check_dim(nrows(a), ncols(a), arr)
  z = ZZMatrix(nrows(a), ncols(a), arr)
  return z
end

function (a::ZZMatrixSpace)(d::ZZRingElem)
  z = ZZMatrix(nrows(a), ncols(a), d)
  return z
end

function (a::ZZMatrixSpace)(d::Integer)
  z = ZZMatrix(nrows(a), ncols(a), flintify(d))
  return z
end
```
Well maybe the last one should be kept as it is a very minor optimization (as it avoids allocating to "convert" `d` to a `ZZRingElem` if it is a bit integer). Though as I said, very few people would go through that interface anyway, so maybe nothing to worry about too much...